rasa/core/policies/rule_policy.py

```diff
@@ -858,43 +874,74 @@ def predict_action_probabilities(
+            tracker, domain, use_text_for_last_user_input=True
+        )
 
         # Rasa Open Source default actions overrule anything. If users want to achieve
         # the same, they need to write a rule or make sure that their loop rejects
         # accordingly.
         default_action_name = self._find_action_from_default_actions(tracker)
-        if default_action_name:
+
+        # text has priority over intents including default,
+        # however loop happy path has priority over rules prediction
+        if default_action_name and not rules_action_name_from_text:
             self._prediction_source = DEFAULT_RULES
             return self._prediction(
                 self._prediction_result(default_action_name, tracker, domain)
             )
 
-        # A loop has priority over any other rule.
+        # A loop has priority over any other rule except defaults.
         # The rules or any other prediction will be applied only if a loop was rejected.
         # If we are in a loop, and the loop didn't run previously or rejected, we can
         # simply force predict the loop.
         loop_happy_path_action_name = self._find_action_from_loop_happy_path(tracker)
         if loop_happy_path_action_name:
             self._prediction_source = LOOP_RULES
+            # this prediction doesn't use user input
+            # and happy user input anyhow should be ignored during featurization
             return self._prediction(
                 self._prediction_result(loop_happy_path_action_name, tracker, domain)
             )
 
-        (
-            rules_action_name,
-            source,
-            returning_from_unhappy_path,
-        ) = self._find_action_from_rules(tracker, domain)
-        # we want to remember the source even if rules didn't predict any action
-        self._prediction_source = source
+        # predict rules from text first
+        if rules_action_name_from_text:
+            policy_events = (
+                [LoopInterrupted(True)] if returning_from_unhappy_path_from_text else []
+            )
+            return self._prediction(
+                self._prediction_result(rules_action_name_from_text, tracker, domain),
+                events=policy_events,
+                is_end_to_end_prediction=True,
+            )
 
+        (
+            rules_action_name_from_intent,
+            # we want to remember the source even if rules didn't predict any action
+            self._prediction_source,
+            returning_from_unhappy_path_from_intent,
+        ) = self._find_action_from_rules(
+            tracker, domain, use_text_for_last_user_input=False
+        )
+        # returning_from_unhappy_path is a negative condition, so `or` should be applied
+        returning_from_unhappy_path = (
+            returning_from_unhappy_path_from_text
+            or returning_from_unhappy_path_from_intent
+        )
         policy_events = [LoopInterrupted(True)] if returning_from_unhappy_path else []
```

Comment:
how about extracting the returning of the `prediction` so we avoid have the logic for the `policy_events` duplicated?

Reply:
I don't understand what do you mean. the logic in line `916` is different

Reply:
The only difference I can see are:
* predicted action name / text
* `is_end_to_end_prediction`

in my opinion it might make sense to extract this in method with 3 parameters 

Consider this optional / non blocking

Reply:
@Ghostvv could you please take this over

Reply:
not sure I understood you correctly, but I created a helper `_prediction_with_unhappy_path`, let me know if that's what you meant